File: sdk_js/types/signedData.ts
import { type } from "arktype";
import type { GetEip712DomainReturnType } from "viem";
import { address, bytes32, bytes, uint256, UINT256_MAX } from "./utils";
import { transferDetail, committedTransferDetail } from "./transferDetail";

/** `publicClient.getEip712Domain` が返す `domain` と同一の形（オフチェーン転送用バンドルに含める） */
export type Eip712SignedDomain = GetEip712DomainReturnType["domain"];

const chainIdAsDomain = type("number").or(uint256);

// viem の getEip712Domain は chainId を Number、salt を bytes32 で返す（型上は bigint も許容）。
// `sign.ts` はゼロ salt を domain から除くため、バンドルに `salt` が無い場合がある。
const eip712SignedDomain = type({
  name: type("string"),
  version: type("string"),
  chainId: chainIdAsDomain,
  verifyingContract: address,
  salt: bytes32.optional(),
});

// TransferWithCommit(address from,address to,address token,address executor,uint256 value,uint256 validAfter,uint256 validBefore,bytes32 commitment)
const transferWithCommit = type({
  domain: eip712SignedDomain,
  from: address,
  to: address,
  token: address,
  value: uint256,
  validAfter: uint256,
  validBefore: uint256,
  commitment: bytes32,
  signature: bytes,
}).filter(
  (v) =>
    v.validAfter <= v.validBefore &&
    v.validAfter >= 0n &&
    v.validBefore <= UINT256_MAX &&
    v.validBefore >= 0n,
);

// UniCommitTransfers(address from,address executor,TransferDetail[] details,uint256 validAfter,uint256 validBefore,bytes32 commitment)
const uniCommitTransfers = type({
  domain: eip712SignedDomain,
  from: address,
  details: transferDetail.array(),
  validAfter: uint256,
  validBefore: uint256,
  commitment: bytes32,
  signature: bytes,
}).filter(
  (v) =>
    v.validAfter <= v.validBefore &&
    v.validAfter >= 0n &&
    v.validBefore <= UINT256_MAX &&
    v.validBefore >= 0n &&
    v.details.length > 0,
);

// BatchTransferWithCommit(address from,address executor,CommittedTransferDetail[] details,uint256 validAfter,uint256 validBefore)
const batchTransferWithCommit = type({
  domain: eip712SignedDomain,
  from: address,
  details: committedTransferDetail.array(),
  validAfter: uint256,
  validBefore: uint256,
  signature: bytes,
}).filter(
  (v) =>
    v.validAfter <= v.validBefore &&
    v.validAfter >= 0n &&
    v.validBefore <= UINT256_MAX &&
    v.validBefore >= 0n &&
    v.details.length > 0,
);

// CancelAuthorization(address authorizer,bytes32 commitment)
const cancelAuthorization = type({
  domain: eip712SignedDomain,
  authorizer: address,
  commitment: bytes32,
  signature: bytes,
});

export const arktype = {
  transferWithCommit,
  uniCommitTransfers,
  batchTransferWithCommit,
  cancelAuthorization,
};

export type SignedTransferWithCommit = Omit<
  typeof transferWithCommit.infer,
  "domain"
> & { domain: Eip712SignedDomain };
export type SignedUniCommitTransfers = Omit<
  typeof uniCommitTransfers.infer,
  "domain"
> & { domain: Eip712SignedDomain };
export type SignedBatchTransferWithCommit = Omit<
  typeof batchTransferWithCommit.infer,
  "domain"
> & { domain: Eip712SignedDomain };
export type SignedCancelAuthorization = Omit<
  typeof cancelAuthorization.infer,
  "domain"
> & { domain: Eip712SignedDomain };

File: sdk_rust/src/types/signed_data.rs
//! オフチェーンで転送する署名済みバンドル。
//!
//! `domain` はオンチェーン `eip712Domain()` を [`crate::sign::domain_for_typed_data_sign`] で正規化したもの（ゼロ `salt` は省略されうる）。トランザクションに載せるのは送金内容と署名であり、**`domain` 本体はオンチェーンには渡さない**（[`crate::send_transaction::signature_transfer`] がエンコードするのはコントラクト関数引数のみ）。

use alloy::primitives::{Address, B256, Bytes, U256};
use alloy::sol_types::Eip712Domain;

use super::transfer_detail::CommittedTransferDetail;

/// 単一送金＋コミットメントの署名結果（型 `TransferWithCommit`）。
#[derive(Clone, Debug)]
pub struct SignedTransferWithCommit {
    /// EIP-712 ドメイン（正規化済み）。
    pub domain: Eip712Domain,
    /// メッセージの `from`。
    pub from: Address,
    pub to: Address,
    pub token: Address,
    pub value: U256,
    pub commitment: B256,
    /// 署名メッセージの `validAfter`（`uint256`）。コントラクトの `validTimestamp` と整合する値（通常はブロックタイムスタンプと比較される単位）。
    pub valid_after: U256,
    /// 署名メッセージの `validBefore`（`uint256`）。上に同じ。
    pub valid_before: U256,
    /// ECDSA 署名（`bytes`）。
    pub signature: Bytes,
}

/// `UniCommitTransfers` の署名結果。
#[derive(Clone, Debug)]
pub struct SignedUniCommitTransfers {
    /// EIP-712 ドメイン。
    pub domain: Eip712Domain,
    pub from: Address,
    /// オフチェーン明細（executor はメッセージ内、`Signed*` には含めない）。
    pub details: Vec<super::transfer_detail::TransferDetail>,
    pub commitment: B256,
    /// [`SignedTransferWithCommit`] の `valid_after` / `valid_before` と同じ意味。
    pub valid_after: U256,
    pub valid_before: U256,
    pub signature: Bytes,
}

/// `BatchTransferWithCommit` の署名結果。
#[derive(Clone, Debug)]
pub struct SignedBatchTransferWithCommit {
    /// EIP-712 ドメイン。
    pub domain: Eip712Domain,
    pub from: Address,
    pub details: Vec<CommittedTransferDetail>,
    /// [`SignedTransferWithCommit`] の `valid_after` / `valid_before` と同じ意味。
    pub valid_after: U256,
    pub valid_before: U256,
    pub signature: Bytes,
}

/// `CancelAuthorization` の署名結果。
#[derive(Clone, Debug)]
pub struct SignedCancelAuthorization {
    /// EIP-712 ドメイン。
    pub domain: Eip712Domain,
    pub authorizer: Address,
    pub commitment: B256,
    pub signature: Bytes,
}
